pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -316,6 +317,229 @@ func (p *staticPolicy) updateCPUsToReuse(pod *v1.Pod, container *v1.Container, c
 	p.cpusToReuse[string(pod.UID)] = p.cpusToReuse[string(pod.UID)].Difference(cset)
 }
 
+// validatePodScopeResources checks for the "empty shared pool" scenario. This occurs
+// when the sum of exclusive container CPU requests consumes the entire pod-level
+// budget, leaving no CPUs for containers that require a shared pool. Such a
+// configuration is invalid because it would lead to containers in the shared
+// pool having an empty cpuset, causing them to run on the node's shared pool
+// and breaking NUMA affinity.
+func (p *staticPolicy) validatePodScopeResources(logger logr.Logger, pod *v1.Pod) error {
+	podTotalCPUs := p.podGuaranteedCPUs(logger, pod)
+
+	// Check for empty shared pool for standard init containers.
+
+	sumOfSidecarExclusiveCPUs := 0
+	maxStandardInitExclusiveCPUs := 0
+	hasSharedStandardInitContainer := false
+	for _, container := range pod.Spec.InitContainers {
+		guaranteedCPUs := p.guaranteedCPUs(logger, pod, &container)
+		if podutil.IsRestartableInitContainer(&container) {
+			sumOfSidecarExclusiveCPUs += guaranteedCPUs
+			continue
+		}
+
+		if guaranteedCPUs > 0 && guaranteedCPUs > maxStandardInitExclusiveCPUs {
+			maxStandardInitExclusiveCPUs = guaranteedCPUs
+		} else {
+			hasSharedStandardInitContainer = true
+		}
+	}
+
+	// This check ensures that the pool of CPUs available to standard init containers
+	// (the total pod budget minus any sidecars) is large enough to satisfy the
+	// request of the largest standard init container that requires exclusive CPUs.
+	// Since standard init containers run sequentially, we only need to fit the largest one.
+	poolForStandardInits := podTotalCPUs - sumOfSidecarExclusiveCPUs
+	if maxStandardInitExclusiveCPUs > poolForStandardInits {
+		return fmt.Errorf("pod rejected, largest exclusive init container requests %d cpus, but only %d available", maxStandardInitExclusiveCPUs, poolForStandardInits)
+	}
+
+	// This check ensures that if there are any standard init containers that need a
+	// shared pool, that pool is not empty. An empty pool would occur if the pod
+	// budget is fully consumed by sidecars.
+	if hasSharedStandardInitContainer && sumOfSidecarExclusiveCPUs >= podTotalCPUs {
+		return fmt.Errorf("pod rejected, pod has shared init containers but no cpus available for them")
+	}
+
+	// Check for empty shared pool for concurrently running containers (app and restartable init containers).
+
+	// Sum CPU requests for all containers that run for the full pod lifetime
+	// (application containers and restartable init containers) to determine the
+	// total exclusive CPU usage that might impact the shared pool.
+	hasSharedAppContainers := false
+	sumOfLongRunningExclusiveCPUs := 0
+	for _, container := range pod.Spec.Containers {
+		guaranteedCPUs := p.guaranteedCPUs(logger, pod, &container)
+		sumOfLongRunningExclusiveCPUs += guaranteedCPUs
+
+		if guaranteedCPUs == 0 {
+			hasSharedAppContainers = true
+		}
+	}
+	for _, container := range pod.Spec.InitContainers {
```

Comment:
nit: just add in `sumOfSidecarExclusiveCPUs` rather than looping over this again.

Reply:
Simplified the func to only have two for loops (standard and init container). Refer to https://github.com/kubernetes/kubernetes/pull/134768#discussion_r2488279698.